sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala

```diff
@@ -52,7 +52,10 @@ case class EnsureRequirements(conf: SQLConf) extends Rule[SparkPlan] {
       case (child, distribution) =>
         val numPartitions = distribution.requiredNumPartitions
           .getOrElse(conf.numShufflePartitions)
-        ShuffleExchangeExec(distribution.createPartitioning(numPartitions), child)
+        // Like optimizer.CollapseRepartition removes adjacent repartition operations,
+        // adjacent repartitions performed by shuffle can be also removed.
+        val newChild = if (child.isInstanceOf[ShuffleExchangeExec]) child.children.head else child
```

Comment:
This reminds me of https://github.com/apache/spark/pull/26946. cc @cloud-fan, @maryannxue and @stczwd FYI

Reply:
To avoid the case @HyukjinKwon pointed out above, it seems we need to check if `outputPartitioning` is the same for narrowing down this optimization scope.

Btw, in my opinion, to avoid complicating the `EnsureRequirements` rule more, it would be better to remove these kinds of redundant shuffles in a new rule after `EnsureRequirements` like https://github.com/apache/spark/pull/27096.

Reply:
>  it seems we need to check if outputPartitioning is the same for narrowing down this optimization scope.

Do you mean we should check whether `outputPartitioning` of `ShuffleExchangeExec` to be inserted and the one of the existing `ShuffleExchangeExec`?
If you mean so, it should already match [this condition](https://github.com/apache/spark/pull/29677/files#diff-cdb577e36041e4a27a605b6b3063fd54R48).

Does it work that just removing the existing `ShuffleExchange` and inserting the new `ShuffleExchange` whose `outputPartitioning` satisfies the required `Distribution`, doesn't it?

> Btw, in my opinion, to avoid complicating the EnsureRequirements rule more, it would be better to remove these kinds of redundant shuffles in a new rule after EnsureRequirements like #27096.

Having a new rule sound better.

Reply:
> Do you mean we should check whether outputPartitioning of ShuffleExchangeExec to be inserted and the one of the existing ShuffleExchangeExec?

I didn't mean so; have you checked #26946? This PR currently removes shuffles incrrectly;
```
scala> spark.range(1).selectExpr("id as a").write.saveAsTable("test")
scala> sql("SELECT /*+ REPARTITION(5) */ * FROM test ORDER BY a").explain()
== Physical Plan ==
*(2) Sort [a#5L ASC NULLS FIRST], true, 0
+- Exchange rangepartitioning(a#5L ASC NULLS FIRST, 200), true, [id=#53]
   +- Exchange RoundRobinPartitioning(5), false, [id=#52] <--- !!! Removed? !!!
      +- *(1) ColumnarToRow
         +- FileScan parquet default.test[a#5L] Batched: true, DataFilters: [], Format: Parquet, Location: InMemoryFileIndex[file:/Users/maropu/Repositories/spark/spark-master/spark-warehouse/test], PartitionFilters: [], PushedFilters: [], ReadSchema: struct<a:bigint>
```

Reply:
I have considered such a case but if a shuffle/partitioning performs after its immediate child of another shuffle/partitioning, is the child shuffle/partitioning  meaningful?
In the example above, is the result not different regardless of `RoundRobinPartitioning` removed or not?

I checked #26946 and I understand that the root cause of that issue was the bug about [how to handle hints in the parser](https://github.com/apache/spark/pull/26946#issuecomment-580861982) so the approach which uses optimization was wrong and incomplete for that issue.

The solutions are similar between this PR and that PR but the issue is difference.
This PR just focuses on removing redundant shuffles.

I might misunderstand about what you point out and that PR, so please let me correct.

Reply:
Ah, I got your point and it makes sense. Coud you update the PR description? The current one only describes the same output partioning cases.

Reply:
Yeah, I've updated. Thanks.